src/lib/Inference.hs

```diff
@@ -302,191 +327,115 @@ checkOrInferRho (WithSrc pos expr) reqTy = do
-      where gn = asGlobal $ varName n
-inferUDecl False (UData      _ _  ) = error "data definitions should be top-level"
-inferUDecl False (UInterface _ _ _) = error "interface definitions should be top-level"
-
-freshClassGenName :: MonadBuilder m => m Name
-freshClassGenName = do
-  scope <- getScope
-  let v' = genFresh (Name TypeClassGenName "classgen" 0) scope
-  builderExtend $ asFst $ v' @> (UnitTy, UnknownBinder)
-  return v'
-
-mkMethod :: UAnnBinder -> UInferM (Label, Type)
-mkMethod (Ignore _) = error "Methods must have names"
-mkMethod (Bind (v:>ty)) = do
-  ty' <- checkUType ty
-  return (nameToLabel v, ty')
-
-mkSuperclass :: UType -> UInferM (Label, Type)
-mkSuperclass ty = do
+  case maybeName of
+    Nothing -> do
+      _ <- withNameHint hint $ emitAnn InstanceLet (Atom instanceDict)
+      return mempty
+      where hint = Name TypeClassGenName "instance" 0
+    Just instanceName -> do
+      instanceVal <- withNameHint instanceName $ emitAnn PlainLet (Atom instanceDict)
+      return $ instanceName @> SubstVal instanceVal
+
+inferDataDef :: UDataDef -> UInferM DataDef
+inferDataDef (UDataDef (tyConName, paramBs) dataCons) =
+  withNestedBinders paramBs \paramBs' -> do
+    dataCons' <- forM dataCons \(dataConName, argBs) ->
+                   withNestedBinders argBs \argBs' ->
+                     return $ DataConDef dataConName argBs'
+    return $ DataDef tyConName paramBs' dataCons'
+
+inferInterfaceDataDef :: SourceName -> Nest UAnnBinder -> [UType] -> [UType] -> UInferM ClassDef
+inferInterfaceDataDef className paramBs superclasses methods =
+  withNestedBinders paramBs \paramBs' -> do
+    superclasses' <- mapM checkUType superclasses
+    methods'     <- mapM checkUType methods
+    let dictContents = PairTy (TupleTy superclasses') (TupleTy methods')
+    return $ DataDef className paramBs'
+               [DataConDef ("Mk"<>className) (Nest (Ignore dictContents) Empty)]
+
+withNestedBinders :: Nest UAnnBinder -> (Nest Binder -> UInferM a) -> UInferM a
+withNestedBinders Empty cont = cont Empty
+withNestedBinders (Nest (UAnnBinder b ty) rest) cont = do
   ty' <- checkUType ty
-  -- TODO: think about the scope of these names
-  l <- freshClassGenName
-  return (nameToLabel l, ty')
-
--- TODO: just make Name and Label the same thing
-nameToLabel :: Name -> Label
-nameToLabel = pprint
-
-mkLabeledItems :: [(Label, a)] -> LabeledItems a
-mkLabeledItems items = foldMap (uncurry labeledSingleton) items
-
-emitConstructors :: DataDef -> UInferM ()
-emitConstructors def@(DataDef tyConName _ dataConDefs) = do
-  let tyConTy = getType $ TypeCon def []
-  checkNotInScope tyConName
-  extendScope $ tyConName @> (tyConTy, DataBoundTypeCon def)
-  forM_ (zip [0..] dataConDefs) \(i, DataConDef dataConName _) -> do
-    let dataConTy = getType $ DataCon def [] i []
-    checkNotInScope dataConName
-    extendScope $ dataConName @> (dataConTy, DataBoundDataCon def i)
-
-emitMethodGetters :: DataDef -> UInferM ()
```

Comment:
How come we got rid of all this? Do all of those live in the renaming pass now?

Reply:
actually, they mostly live in Builder now